### ISSUE TITLE
Reproducing ours-sync

### ISSUE THREAD

**Issue by josejhlee** [AniSundar18/AlignedForensics#3]
Hi authors,

Thank you for your work in this field. I was trying to reproduce "ours-sync"(ICLR25) training with 20k images (since the performance saturated at there), and I was unsuccessful. 

1. Could you share the bash script or the hyperparameters for it? 
2. Did you use LSUN data for the 20k training?
3. Is LSUN critical for generalization?

**Comment by AniSundar18** [comment]
Thanks for sharing your code and walking me through the issue.

The main discrepancy seems to come from comparing models trained on 10k images per class with the provided checkpoints, which were trained on the full dataset (~170k per class). This difference is expected. For a fair comparison with a 10k/10k setting, please refer to Table 3 in Section 5.4 of the paper. (Note: in that table, fake images are not resized to ignore a spurious correlation in the main baseline, as discussed in section 5.4 .)

When I trained a model using your codebase, I noticed a few differences from our setup:

1. Training split – Your code splits 10k images into train/val, so only 8k per class are used for training. In our experiments, we train on 10k images per class and use the fixed validation set from the DMIDetection repo across all runs for consistency. This doesn’t seem to significantly affect results.
2. Batch size – We used a batch size of 64 (effectively 128 with reconstructions) in the paper. This will be updated in the script, but it also doesn’t appear to cause large performance changes.

When using your setup, validation accuracy is around 85%, whereas we see about 75% using the full validation set (our setup)—this gap doesn’t seem to cause major issues in downstream conclusions. We show the results below (we report accuracy with a 0.5 threshold), for these results, we use both real and fake images from the evaluation dataset which we created for our ICML paper, both real and fake images have both clean and post-processed images (including resizing), distinguishing it from Table 3 (Section 5.4),

<img width="634" height="90" alt="Image" src="https://github.com/user-attachments/assets/fda2aab7-5910-4aaa-a743-c731770c9b23" />

The first two rows show our reproduction of the experiment using the code that we have shared. The third row corresponds to running the code you shared, and the final row shows one of the weight sets we used to report results in the ICLR paper (Section 5.4). Please let me know if these results differ from what you observe with your method.

Regarding Stay-Positive:

1. We start from a pre-trained detector, freeze the backbone, and retrain only the final layer (removing dropout in that layer).
2. The update follows Algorithm 1 from the paper: after a standard gradient step, we zero-out any weights which are negative.
3. Learning rate, dataset, and training recipe are the same as first-stage training, except batch size is 1024.
4. Batch norm mean/variance continue updating, but scale/shift parameters are fixed. This choice is not very significant, performance stays the similar irrespective of this choice.

In general, we recommend starting Stay-Positive from the strongest possible base model, ideally trained on the full dataset, so the detector associates some patterns with fake images as well (see Appendix A.6). 

Hope this helps clarify the differences, please reply in case there are further issues!

**Comment by josejhlee** [comment]
Thank you again for your incredibly helpful and detailed reply. It has been invaluable for our training runs.

Following your suggestions, we have conducted bigger scale experiments to align our setup as closely as possible with yours. Specifically:
- We trained the Ours-Sync model on the full dataset of 180k images (360k in total, including real and fake), using the exact indices from MS COCO and LSUN that Corvi et al. (2022) used. (https://github.com/grip-unina/DMimageDetection) We adopted a 90/10 split, resulting in ~168k training images. 
- To eliminate any potential variance, we used your provided scripts to generate the reconstructed images.
- We tested with batch sizes of both 64 and 128.

This has definitely helped close the gap somewhat, and our results have improved. However, we're still observing a significant performance gap between our trained models and your provided Ours-Sync checkpoint on the test sets.
During our full-dataset training runs, our validation accuracy consistently reaches about 91.6% on bs 64 and 92.3% on bs 128. 
To resolve this, would it be possible for you to share the exact data you have used in your training (in ours-sync and staypos) that is usable right away with the codebase? I understand that there shouldn’t be many ways the re-implementation could have gone wrong, however at this point, I really want to nail down and align on every step there exists...

And once more, thank you for your time and guidance!

**Comment by AniSundar18** [comment]
Hi, 

I took some time to go through your codebase (Gloriel621). There seems to be a discrepancy in terms of how the validation set is handled. I ran the experiment with a val split of 0.1 as you had suggested. I also tried using a batch size of 64 (effectively 128 with corresponding inversions). The training script printed the following, 

# training batches = 2521
# validation batches = 5602

There are less than half as many training pairs as there are validation data instances. This translates to 322,688 training instances (2521*64*2), which is roughly 90% of the training data. However, the validation dataset contains 5602 batches which consists of 358,528 (5602*64 since no sync batches for validation) images, which is the size of the whole training dataset. I looked at the code (attaching below),

```
def create_dataloader(opt, subdir=".", is_train=True):

    np.random.seed(opt.seed)

    if getattr(opt, "real_dir", None) and getattr(opt, "fake_dir", None):

        if not hasattr(opt, "_tfds"):
            tfm = make_processing(opt)

            ds_train = CapDataset_(
                opt.real_dir, opt.fake_dir,
                transform=tfm,
                batched_syncing=opt.batched_syncing,
                data_cap=opt.data_cap,
                use_inversions=opt.use_inversions,
                seed=opt.seed,
            )

            val_cap = None
            if opt.data_cap is not None:
                val_cap = max(1, opt.data_cap // 5)

             # ==========================
            ds_valid = CapDataset_(
                opt.real_dir, opt.fake_dir,
                transform=tfm,
                batched_syncing=False,        # flat batches
                data_cap=val_cap,             # smaller set
                use_inversions=opt.use_inversions,
                seed=opt.seed,
            )

            **_val_len   = int(len(ds_train) * opt.val_split)
            train_len = len(ds_train) - val_len
            gen = torch.Generator().manual_seed(opt.seed)
            idx_train, _ = torch.utils.data.random_split(
                range(len(ds_train)), [train_len, val_len], generator=gen
            )

            opt._tfds = {
                "train": Subset(ds_train, idx_train.indices),
                "valid": ds_valid,            # ← no Subset here
            }_**
             # ==========================

        dataset = opt._tfds["train" if is_train else "valid"]
    else:
        dataroot = os.path.join(opt.dataroot, subdir)
        if (opt.data_cap is not None or opt.batched_syncing) and is_train:
            tfm = make_processing(opt)
            dataset = CapDataset(root_dir=dataroot, data_cap=opt.data_cap, transform=tfm, batched_syncing=opt.batched_syncing, use_inversions=opt.use_inversions, seed=opt.seed)
            dataset = torch.utils.data.ConcatDataset([dataset])
        else:
            dataset = get_dataset(opt, dataroot)
    loader = DataLoader(dataset, batch_size=opt.batch_size, sampler=get_bal_sampler(dataset) if is_train else None, num_workers=int(opt.num_threads))
    return loader
```

I have highlighted the part of this code that does the train, val split. ds_valid is pretty much the same as ds_train. The random split only selects the indices for the training data (90% of total) and leaves the val data as it is. Ideally there should be Subset(ds_valid, idx_valid.indices) as well. 

As a result the validation set contains every single training example (and a little more). This causes the best model to be the one which has overfitted the training data the most. I am currently running an experiment with a fixed version of the same to check for further issues. 

Now, about the validation set. We use the same validation images as provided by the following codebase (https://github.com/grip-unina/DMimageDetection). They provide their data, including the splits here (https://www.grip.unina.it/download/prog/DMimageDetection/latent_diffusion_trainingset.zip). Upon downloading, there will be a folder called valid with validation data, please use that as your validation images. I hope that this helps with reproducibility, however, please follow up if there are any further issues, Thanks!
